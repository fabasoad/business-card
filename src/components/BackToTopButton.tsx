import * as React from 'react'

export default function BackToTopButton() {
  const [isVisible, setIsVisible] = React.useState<boolean>(false)

  window.addEventListener(
    'scroll',
    () => setIsVisible(window.scrollY > 300),
    { passive: true }
  )

  const handleClick = (e) => {
    e.preventDefault()
    window.scrollTo({
      top: 0,
      left: 0,
      behavior: 'smooth'
    })
  }

  return (
    <a onClick={handleClick} className={(isVisible ? 'show ' : '') + 'back-to-top'}></a>
  )
}
