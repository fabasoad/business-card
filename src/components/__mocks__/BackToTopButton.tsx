import * as React from 'react'

export default function BackToTopButton() {
  return (
    <div data-testid="BackToTopButton"></div>
  )
}
