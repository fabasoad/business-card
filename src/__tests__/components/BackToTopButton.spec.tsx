/// <reference types="jest" />
import * as React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import BackToTopButton from '../../components/BackToTopButton'

let wrapper: ShallowWrapper
let windowScrollSpy: jest.SpyInstance<void, [number, number]>

beforeAll(() => {
  wrapper = shallow(<BackToTopButton />)
  windowScrollSpy = jest.spyOn(window, 'scrollTo').mockImplementation()
})

test('should scroll to top', () => {
  const preventDefaultSpy = jest.fn()
  wrapper.find('a').simulate('click', {
    preventDefault: preventDefaultSpy
  })
  expect(preventDefaultSpy).toHaveBeenCalledTimes(1)
  expect(windowScrollSpy).toHaveBeenCalledWith({
    top: 0,
    left: 0,
    behavior: 'smooth'
  })
})

new Array<number>(300, 301).forEach((v) => {
  test(`should render BackToTopButton correctly as ${v === 300 ? 'not ' : ''}visible`, () => {
    Object.defineProperty(window, 'scrollY', { value: v })
    window.dispatchEvent(new Event('scroll'))
    expect(wrapper).toMatchSnapshot()
  })
})
