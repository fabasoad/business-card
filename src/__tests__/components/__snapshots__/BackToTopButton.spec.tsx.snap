// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BackToTopButton correctly as not visible 1`] = `
<a
  className="back-to-top"
  onClick={[Function]}
/>
`;

exports[`should render BackToTopButton correctly as visible 1`] = `
<a
  className="show back-to-top"
  onClick={[Function]}
/>
`;
